### PR TITLE
ENG-1679 Add cypress testing to app-builder

### PR DESCRIPTION
**Pull request by ichalagashvili**

### CODE REVIEW

**Comment by entando-jx** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ichalagashvili**

If they are not already assigned, you can assign the PR to them by writing `/assign @ichalagashvili` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=entando%2Fapp-builder).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/entando/app-builder/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: test/cypress/support/utils.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line import/prefer-default-export
+export const generateRandomId = () => Math.floor(Math.random() * Math.floor(999999999999999));
```
Comment (avdev4j) [inline test/cypress/support/utils.js:2]: This should avoid having two pages with the same id but for sure it's not the best way to do it...

File: test/cypress/support/exception-handler.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// returning false here prevents Cypress from failing the test
+Cypress.on('uncaught:exception', () => false);
```
Comment (avdev4j) [inline test/cypress/support/exception-handler.js:2]: With this Cypress will not fail if there is an error in the console. Maybe we should reconsider it.

File: test/cypress/integration/roles.spec.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+      cy.get('[data-cy=save-role').contains('Save').click();
+      cy.get('[data-cy=user-roles]').contains('Roles').should('be.visible');
+    });
+  });
+
+
+  describe('Edit an existing role', () => {
+    it('Should edit an existing role', () => {
+      cy.visit(`/role/edit/${roleCode}`);
```
Comment (avdev4j) [inline test/cypress/integration/roles.spec.js:38]: I wonder if we can think of adding this kind of thing in the support part,
Here is an example: https://github.com/jhipster/generator-jhipster/tree/main/generators/cypress/templates/src/test/javascript/cypress/support

Even if sometimes I find it "too much"

File: package.json [hunk 4]
```diff
@@ -183,6 +188,7 @@
     "@testing-library/react": "^10.4.9",
     "@testing-library/user-event": "^12.2.0",
     "commander": "^3.0.2",
+    "cypress": "4.10.0",
```
Comment (avdev4j) [inline package.json:191]: Could you try with the last version -> 6.4.0?

File: package.json [hunk 5]
```diff
@@ -99,6 +99,10 @@
     "start": "node scripts/start.js",
     "build": "node scripts/build.js",
     "test": "node scripts/test.js --env=jsdom",
+    "cypress": "cypress open",
+    "cypress-test": "cross-env NODE_ENV=test cypress open",
```
Comment (avdev4j) [inline package.json:103]: We can remove this one

**Review by avdev4j — CHANGES_REQUESTED** [review]

**Review by gidesan — CHANGES_REQUESTED** [review]
On top of the CR by Anthony, I see there is usage of both `dataCy` and `data-cy`.
I'd adopt something more tech-agnostic like `data-testid` which is also the one used by react testing library.

Last but not least: what about including a library to cleanup testids like https://www.npmjs.com/package/babel-plugin-jsx-remove-data-test-id ? @avdev4j

**Comment by avdev4j** [comment]
It makes sense if we can having only one specific attribute for each kind of test.

About cleaning them, I would say that this should not be a problem while we do not suffer from a slower build.

File: package.json (hunk 4 above)
Comment (ichalagashvili) [reply package.json:191]: does not work :(

File: test/cypress/support/exception-handler.js (hunk 2 above)
Comment (ichalagashvili) [reply test/cypress/support/exception-handler.js:2]: we have some errors atm in console so we need to keep this as it is for now until we fix all errors

Comment (avdev4j) [reply test/cypress/support/exception-handler.js:2]: sounds good.

**Review by gidesan — APPROVED** [review]

**Comment by entando-jx** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/entando/app-builder/pull/991#pullrequestreview-582842425" title="Approved">gidesan</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=entando%2Fapp-builder).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/entando/app-builder/blob/master/OWNERS)~~ [gidesan]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**